[master]Add Ballerina Deployment Guide (#9660)
ballerina-platform/ballerina-dev-website: swan-lake/development-tutorials/deployment-guide/k8s-deployment.md

> ---

Reviewer: @sm1990 Sometimes I am getting the following error when trying to render this page locally. Any idea?

<img width="1034" alt="image" src="https://github.com/user-attachments/assets/30d818ed-a18b-4323-b6a0-ed8f8be9f1be" />

}
```
This code is deployment-agnostic and only handles the business logic. Ballerina automatically builds and generates the Kubernetes deployment artifacts.

> **Note:** If you are using a Ballerina cache in a Kubernetes environment, note that the cache is local to each pod. 
> Consider using a Redis server to manage the cache if your application requires a distributed caching solution.

### Overriding default Kubernetes configurations

Create a `Cloud.toml` file in the project directory to customize the deployment parameters. This file will allow you to override the compiler's default values.

**Sample Cloud.toml**

```toml
[container.image]
repository="wso2inc"  # Docker hub repository name. Default Value: ""
name="hello" # container name. Default Value: $USER-$MODULE_NAME
tag="v0.1.0" # Tag of the container. Default Value: "latest"
```
See [Ballerina Code to Cloud Specification](https://github.com/ballerina-platform/ballerina-spec/blob/master/c2c/code-to-cloud-spec.md#containerimage) for more information.

### Building the Kubernetes artifacts

Use the `bal build` command to build the Ballerina package. This will generate Kubernetes YAML files and Docker container artifacts.

> >**Note:** For macOS users with Apple Silicon chips, it is necessary to set the environment variable DOCKER_DEFAULT_PLATFORM to linux/amd64 before building the image. This requirement exists because the Ballerina Docker image does not currently support Apple Silicon chips.

Reviewer: ```suggestion
>**Note:** For macOS users with Apple Silicon chips, it is necessary to set the environment variable `DOCKER_DEFAULT_PLATFORM` to `linux/amd64` before building the image. This requirement exists because the Ballerina Docker image does not currently support Apple Silicon chips.
```

The artifacts are located in `target/kubernetes/hello`.

### Deploying the application to Kubernetes

Push the Docker image to Docker Hub.
```
$ docker push wso2inc/hello:v0.1.0
```

Deploy the generated Kubernetes artifacts.
```
$ kubectl apply -f /path/to/target/kubernetes/hello
```

Expose the service via NodePort for local testing.

```
$ kubectl expose deployment hello-deployment --type=NodePort --name=hello-svc-local
service/hello-svc-local exposed
```

Get the Kubernetes service details.
```
$ kubectl get svc

NAME                        TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)          AGE
hello-svc               ClusterIP       10.97.140.84        <none>        9090/TCP         13m
hello-svc-local         NodePort        10.108.87.21        <none>        9090:31360/TCP   2m18s
```
Access the service using the external IP and NodePort

For Minikube users, you can obtain the IP using.
```yaml
$ minikube ip
```
> **Tip:** you can access your service locally via `kubectl port-forward`
 `$ kubectl port-forward svc/hello-k8s-svc  8081:8080`

> Test the service using curl

Reviewer: ```suggestion
Test the service using curl.
```

```
$ curl http://<EXTERNAL-IP>:<NODE-PORT>/helloWorld/greeting
```

## Kubernetes deployment with config maps and autoscaling

For a more complex deployment, this demonstrates a Ballerina app that reads a greeting from a config map and responds to HTTP requests. It covers Kubernetes deployment with resource limits, autoscaling, config maps, and health probes using Code to Cloud. 

1. Create a new Ballerina package.

> $ bal new hello_k8s

Reviewer: Missing ``` for code segment?

Reviewer 2: Fixed

       kubectl apply -f /home/wso2/user/hello_k8s/target/kubernetes/hello_k8s

       Execute the below command to access service via NodePort: 
       kubectl expose deployment hello-k8s-deployment --type=NodePort --name=hello-k8s-svc-local

       target/bin/hello_k8s.jar
   ```
> **Tip:** You can modify the Ballerina.toml file to include a build option, allowing the integration of the `bal build` command with specific configurations.
> ```
> [build-options] 
>  cloud = "k8s"
> ```

> **Tip:** You can build the GraalVM executable in Kubernetes.
> 
>  GraalVM enhances Kubernetes deployments of Ballerina applications by enabling faster startup times and reduced memory usage. These advantages make it ideal for microservices workloads that demand rapid scaling and minimal resource consumption in a containerized environment. GraalVM’s native image feature uses ahead-of-time compilation, producing lightweight containers and improving Kubernetes deployment efficiency. The GraalVM build is best suited for deployments that prioritize fast scaling and reduced resource usage, while the standard JVM build is more appropriate for services that do not require such optimizations.

> >  Execute the command below to build a Ballerina package.

Reviewer 2: Let's add a line saying the build time will be higher for Graal VM builds.

7. Push Docker image to Docker Hub.
```
 $ docker push wso2inc/hello-k8s:v0.1.0
```

8. Deploy the Kubernetes artifacts. 
```
 $ kubectl apply -f target/kubernetes/hello_k8s
```
9. Testing the Kubernetes Deployment.
   Verifying the Kubernetes deployment ensures that the application is running correctly and meets the expected functionality within the Kubernetes environment.
  - Verify the pods
    ```
    $ kubectl get pods
    ```
  - Expose the service via NodePort
    ```
    $ kubectl expose deployment hello-k8s-deployment --type=NodePort --name=hello-k8s-svc-local exposed
    ```
  - Get the External IP and port
    ```
    $ kubectl get svc
    NAME                      TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)          AGE
    hello-k8s-svc         ClusterIP   10.96.173.207   <none>        9090/TCP,9091/TCP               5m11s
    hello-k8s-svc-local   NodePort    10.99.245.41    <none>        9090:30342/TCP,9091:30515/TCP   66s
    kubernetes            ClusterIP   10.96.0.1       <none>        443/TCP                         130m
    ```

>   - Execute the curl http://192.168.49.2:30342/helloWorld/greeting command to access the deployed service via cURL.

Reviewer: ```suggestion
  - Execute the following command to access the deployed service via cURL.
```

        env:
          KUBE_CONFIG_DATA: ${{ secrets.KUBE_CONFIG_DATA }}
        run: |
          echo "${KUBE_CONFIG_DATA}" | base64 --decode > $HOME/.kube/config
          kubectl apply -f <path to generated artifacts>
```
The push event can initiate a workflow to the deployment process to the Kubernetes cluster.
- **Step 02:** This step builds the Ballerina project, compiling the source code and preparing the necessary artifacts for deployment in Kubernetes.
- **Step 03:** This step logs into a Docker registry (such as Docker Hub or a private registry) to allow pushing the Docker image created in the build to the registry.
- **Step 04:** Pushes the Docker image to the specified Docker registry.
- **Step 05:** This step installs kubectl, the Kubernetes command-line tool, which is necessary for managing Kubernetes clusters and deploying applications.
- **Step 06:** Deploy the application to a Kubernetes cluster by applying the Kubernetes YAML configuration files generated by the Ballerina build in Step 2.
>**Note:** Replace <path to generated artifacts> with the actual path to the Kubernetes YAML files generated by the Ballerina build command (e.g., target/kubernetes).

> See [Code to Cloud deployment](https://ballerina.io/learn/code-to-cloud-deployment/) for more information on deployment options in Kubernetes and Docker.

Reviewer: Are we going to have https://ballerina.io/learn/code-to-cloud-deployment/ pages as well after these changes?

Reviewer: Shouldn't we reword this to `serverless deployment`?

### Deploying the application to Kubernetes

Push the Docker image to Docker Hub.
```
$ docker push wso2inc/hello:v0.1.0
```

Deploy the generated Kubernetes artifacts.
```
$ kubectl apply -f /path/to/target/kubernetes/hello
```

Expose the service via NodePort for local testing.

```
$ kubectl expose deployment hello-deployment --type=NodePort --name=hello-svc-local
service/hello-svc-local exposed
```

Get the Kubernetes service details.
```
$ kubectl get svc

NAME                        TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)          AGE
hello-svc               ClusterIP       10.97.140.84        <none>        9090/TCP         13m
hello-svc-local         NodePort        10.108.87.21        <none>        9090:31360/TCP   2m18s
```
Access the service using the external IP and NodePort

For Minikube users, you can obtain the IP using.
```shell
$ minikube ip
```
> **Tip:** you can access your service locally via `kubectl port-forward`
 `$ kubectl port-forward svc/hello-k8s-svc  8081:8080`

Test the service using cURL.

```
$ curl http://<EXTERNAL-IP>:<NODE-PORT>/helloWorld/greeting
```

## Kubernetes deployment with config maps and autoscaling

> For a more complex deployment, this demonstrates a Ballerina app that reads a greeting from a config map and responds to HTTP requests. It covers Kubernetes deployment with resource limits, autoscaling, config maps, and health probes using Code to Cloud. 

Reviewer: Is it clear to say Code to Cloud here?

1. Create a new Ballerina package.

> ```bash

Reviewer 3: ```suggestion
```
```
